### DISCUSSION TITLE
Does keymapper support touchpads or more specifically Apple's builtin trackpad?

### DISCUSSION THREAD

**Discussion by prinzdezibel** [houmain/keymapper#186] [Q&A]
When grabbing the touchpad I'm able to remap click events sucessfully. But unfortunately I'm no longer able to move the cursor arrow with the touchpad. Also gestures like two finger scrolling wouldn't work any longer. The "ButtonLeft" event does work though and can be remapped. Any idea on this?

**Comment by houmain** [comment]
> When grabbing the touchpad I'm able to remap click events sucessfully

Are you using the Linux version? Mapping mice on MacOS is not supported yet (sorry, I should have documented that).
If yes, then could you please show the output of `sudo evtest` for this device.

**Reply by prinzdezibel** [reply]
Sorry, forgot to mention that I'm using Linux on Apple hardware. 

/dev/input/event0:      Apple SMC power/lid events
/dev/input/event1:      Apple Internal Keyboard / Trackpad
/dev/input/event2:      Apple Internal Keyboard / Trackpad
/dev/input/event3:      MacBook Pro J316 Headphone Jack
/dev/input/event4:      Apple Inc. Magic Keyboard with Numeric Keypad
/dev/input/event5:      Apple Inc. Magic Keyboard with Numeric Keypad
/dev/input/event6:      Apple Inc. Magic Trackpad
/dev/input/event7:      Apple Inc. Magic Trackpad
/dev/input/event8:      Keymapper
/dev/input/event9:      Apple Inc. Mouse


I'm interested in getting the builtin trackpad to work, regardless the discussion's title where I was talking about the "Magic Trackpad", which is a separate standalone device. Going to change the title.
 
Output for builtin Trackpad  (/dev/input/event1)


```
Input driver version is 1.0.1
Input device ID: bus 0x1c vendor 0x5ac product 0x343 version 0x352
Input device name: "Apple Internal Keyboard / Trackpad"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 272 (BTN_LEFT)
    Event code 325 (BTN_TOOL_FINGER)
    Event code 328 (BTN_TOOL_QUINTTAP)
    Event code 330 (BTN_TOUCH)
    Event code 333 (BTN_TOOL_DOUBLETAP)
    Event code 334 (BTN_TOOL_TRIPLETAP)
    Event code 335 (BTN_TOOL_QUADTAP)
  Event type 3 (EV_ABS)
    Event code 0 (ABS_X)
      Value      0
      Min    -7456
      Max     7976
      Resolution      97
    Event code 1 (ABS_Y)
      Value      0
      Min    -9283
      Max      163
      Resolution      96
    Event code 24 (ABS_PRESSURE)
      Value      0
      Min        0
      Max     6000
      Resolution       1
    Event code 47 (ABS_MT_SLOT)
      Value      0
      Min        0
      Max       15
    Event code 48 (ABS_MT_TOUCH_MAJOR)
      Value      0
      Min        0
      Max     5000
    Event code 49 (ABS_MT_TOUCH_MINOR)
      Value      0
      Min        0
      Max     5000
    Event code 50 (ABS_MT_WIDTH_MAJOR)
      Value      0
      Min        0
      Max     5000
    Event code 51 (ABS_MT_WIDTH_MINOR)
      Value      0
      Min        0
      Max     5000
    Event code 52 (ABS_MT_ORIENTATION)
      Value      0
      Min   -16384
      Max    16384
    Event code 53 (ABS_MT_POSITION_X)
      Value      0
      Min    -7456
      Max     7976
      Resolution      97
    Event code 54 (ABS_MT_POSITION_Y)
      Value      0
      Min    -9283
      Max      163
      Resolution      96
    Event code 57 (ABS_MT_TRACKING_ID)
      Value      0
      Min        0
      Max    65535
    Event code 58 (ABS_MT_PRESSURE)
      Value      0
      Min        0
      Max     6000
      Resolution       1
  Event type 4 (EV_MSC)
    Event code 4 (MSC_SCAN)
Properties:
  Property type 0 (INPUT_PROP_POINTER)
  Property type 2 (INPUT_PROP_BUTTONPAD)
Testing ... (interrupt to exit)
```

**Reply by houmain** [reply]
Thanks!
Keymapper's virtual device would have to announce these absolute axes, so it can forward the input. Likely these Min/Max/Resolution values should also match the physical device's. I am thinking about creating the virtual device(s) dynamically, depending on the grabbed devices. But this might take a while.
